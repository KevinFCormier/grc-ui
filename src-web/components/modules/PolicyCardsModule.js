/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import { Select, SelectItem } from 'carbon-components-react'
import resources from '../../../lib/shared/resources'
import msgs from '../../../nls/platform.properties'
import _ from 'lodash'

resources(() => {
  require('../../../scss/module-policy-cards.scss')
})

const PolicyCardsSelections = Object.freeze({
  categories: 'categories',
  standards: 'standards',
})

export default class PolicyCardsModule extends React.Component {

  constructor (props) {
    super(props)
    const {viewState: {policyCardChoice=PolicyCardsSelections.categories}} = props
    this.state = {
      policyCardChoice
    }
    this.onChange = this.onChange.bind(this)
  }

  render() {
    const cardData = this.getCardData()
    return (
      <div className='module-policy-cards'>
        {this.renderHeader()}
        {this.renderCards(cardData)}
      </div>
    )
  }

  renderHeader() {
    const { locale } = this.context
    const { policyCardChoice } = this.state
    const choices = this.getPolicyCardChoices(locale)
    const title = msgs.get('overview.policy.overview.title', locale)
    return (
      <div className='header-container'>
        <div className='header-title'>{title}</div>
        <Select id={'select-policy-card-type'} className='header-selection'
          hideLabel={true}
          value={policyCardChoice}
          onChange={this.onChange}>
          {choices.map(({text, value})=> {
            return (
              <SelectItem key={value} text={text} value={value} />
            )
          })}
        </Select>
      </div>
    )
  }

  renderCards(cardData) {
    const { locale } = this.context
    const { handleDrillDownClick } = this.props
    return (
      <div className='card-container-container' >
        {cardData.map((data) => {
          return <PolicyCard key={data.name} data={data} locale={locale} handleClick={handleDrillDownClick} />
        })}
      </div>
    )
  }

  getCardData = () => {
    const { locale } = this.context
    const { policies, activeFilters } = this.props
    const { policyCardChoice } = this.state
    const other = msgs.get('overview.policy.overview.other', locale)

    // loop thru policies
    const dataMap = {}
    const policyMap = {}
    const clusterMap = {}
    policies.map(policy => {
      // get a policy's standards/categories
      let types, key
      const annotations = _.get(policy, 'metadata.annotations', {}) || {}
      switch (policyCardChoice) {
      case PolicyCardsSelections.categories:
        types = annotations['policy.mcm.ibm.com/categories'] || ''
        key = 'categories'
        break
      case PolicyCardsSelections.standards:
        types = annotations['policy.mcm.ibm.com/standards'] || ''
        key = 'standards'
        break
      }
      // backward compatible and if user doesn't supply an annotation
      if (types.length===0) {
        types=other
      }
      types.split(',').forEach(type=>{
        type = type.trim()
        if (type) {
          let name = type
          if (policyCardChoice===PolicyCardsSelections.categories) {
            name = _.capitalize(_.startCase(name))
          }
          const filtered = activeFilters[key] &&  activeFilters[key].size>0 && !activeFilters[key].has(name)
          if (!filtered) {
            let data = dataMap[type]
            if (!data) {
              data = dataMap[type]= {
                name,
                violations: 0,
                counts: {
                  cluster: {
                    violations:0,
                    total:0,
                  },
                  policy: {
                    violations:0,
                    total:0,
                  },
                }
              }
            }

            // if policy has violation
            policyMap[type] = policyMap[type] || {}
            clusterMap[type] = clusterMap[type] || {}
            const statuses = _.get(policy, 'raw.status.status', {})
            const policyName = _.get(policy, 'metadata.name')
            Object.keys(statuses).forEach(clusterName=>{
              const noncompliant = statuses[clusterName].compliant!=='Compliant'
              policyMap[type][policyName] = policyMap[type][policyName] || noncompliant
              clusterMap[type][clusterName] = clusterMap[type][clusterName] || noncompliant
            })
          }
        }
      })
    })

    // tabulate for category or standard
    // by policies
    Object.keys(policyMap).forEach(type=>{
      const data = dataMap[type]
      const policies = policyMap[type]
      Object.keys(policies).forEach(policy=>{

        // this is a policy with this category/standard
        data.counts.policy.total++

        // this policy with this category/standard has a violation
        if (policies[policy]) {
          data.counts.policy.violations++
          data.violations++
        }
      })
    })

    // by clusters
    Object.keys(clusterMap).forEach(type=>{
      const data = dataMap[type]
      const clusters = clusterMap[type]
      Object.keys(clusters).forEach(cluster=>{

        // this is a cluster with this category/standard
        data.counts.cluster.total++

        // this cluster with this category/standard has a violation
        if (clusters[cluster]) {
          data.counts.cluster.violations++
          data.violations++
        }
      })
    })

    // convert to array and sort
    return Object.keys(dataMap).map(key=>{
      return {...dataMap[key]}
    }).sort(({name:an, violations:av}, {name:bn, violations:bv})=>{
      const v = bv-av
      if (v===0) {
        if (an===other && bn!==other) {
          return 1
        } else  if (an!==other && bn===other) {
          return -1
        }
        return an.localeCompare(bn)
      }
      return v
    })
  }

  getPolicyCardChoices = (locale) => {
    if (!this.policyCardChoices) {
      this.policyCardChoices = [
        {
          value: PolicyCardsSelections.categories,
          text: msgs.get('overview.policy.cards.categories', locale),
        },
        {
          value: PolicyCardsSelections.standards,
          text: msgs.get('overview.policy.cards.standards', locale),
        },
      ]
    }
    return this.policyCardChoices
  }

  onChange = (e) => {
    const {value} = this.policyCardChoices[e.currentTarget.selectedIndex]
    this.props.handleDisplayChange(value)
    this.props.updateViewState({policyCardChoice: value})
    this.setState(()=>{
      return {policyCardChoice: value}
    })
  }
}




// functional card component
const PolicyCard = ({data, locale, handleClick}) => {
  const { counts, name } = data
  const countData = Object.keys(counts).map(type=>{
    return {
      ...counts[type],
      violationType: msgs.get(`overview.${type}.violations`, locale),
      type: type
    }
  })
  return (
    <div key={name}>
      <div className='card-container'>
        <div className='card-content'>
          <div className='card-name'>
            {name}
          </div>
          <div className='card-count-content'>
            {countData.map(({violations, total, violationType, type}) => {
              const classes = classNames({
                'card-count-violations': true,
                'violated': violations > 0,
              })
              return (
                //eslint-disable-next-line
                <div key={violationType} className={classes !== 'card-count-violations' ? 'card-count-container' : 'card-count-container violated'} onClick={handleClick(name, type, classes)}>
                  <div className='card-count'>
                    <div className={classes}>
                      {violations}
                    </div>
                    <div className='card-count-total'>
                      {`/${total}`}
                    </div>
                  </div>
                  <div className='card-violation-type'>
                    {violationType}
                  </div>
                </div>
              )
            })}
          </div>
        </div>
      </div>
    </div>
  )
}

PolicyCard.propTypes = {
  data: PropTypes.object,
  handleClick: PropTypes.func,
  locale: PropTypes.string
}


PolicyCardsModule.propTypes = {
  activeFilters: PropTypes.object,
  handleDisplayChange: PropTypes.func,
  handleDrillDownClick: PropTypes.func,
  policies: PropTypes.array,
  updateViewState: PropTypes.func,
  viewState: PropTypes.object,
}
