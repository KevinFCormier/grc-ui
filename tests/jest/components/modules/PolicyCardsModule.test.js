/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import { BrowserRouter } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { policiesTestingDataSet1, policiesTestingDataSet2 } from './ModuleTestingData'
import PolicyCardsModule from '../../../../src-web/components/modules/PolicyCardsModule'

describe('PolicyCardsModule categories view', () => {
  const viewState = {viewState: {policyCardChoice:'categories'}}
  const activeFilters = {activeFilters:{}}
  const updateViewState = jest.fn()
  const handleDrillDownClick = jest.fn()
  const filteredPolicies = policiesTestingDataSet1
  it('renders expand as expected', () => {
    const component = renderer.create(<BrowserRouter><PolicyCardsModule  viewState={viewState}
      updateViewState={updateViewState}
      policies={filteredPolicies}
      activeFilters={activeFilters}
      showPolicyCard={true}
      handleDrillDownClick={handleDrillDownClick} /></BrowserRouter>)
    expect(component).toMatchSnapshot()
  })

  it('renders collapse as expected', () => {
    const component = renderer.create(<BrowserRouter><PolicyCardsModule  viewState={viewState}
      updateViewState={updateViewState}
      policies={filteredPolicies}
      activeFilters={activeFilters}
      showPolicyCard={false}
      handleDrillDownClick={handleDrillDownClick} /></BrowserRouter>)
    expect(component).toMatchSnapshot()
  })
})

describe('PolicyCardsModule standards view', () => {
  const viewState = {viewState: {policyCardChoice:'standards'}}
  const activeFilters = {activeFilters:{}}
  const updateViewState = jest.fn()
  const handleDrillDownClick = jest.fn()
  const filteredPolicies = policiesTestingDataSet2
  it('renders expand as expected', () => {
    const component = renderer.create(<BrowserRouter><PolicyCardsModule  viewState={viewState}
      updateViewState={updateViewState}
      policies={filteredPolicies}
      activeFilters={activeFilters}
      showPolicyCard={true}
      handleDrillDownClick={handleDrillDownClick} /></BrowserRouter>)
    expect(component).toMatchSnapshot()
  })

  it('renders collapse as expected', () => {
    const component = renderer.create(<BrowserRouter><PolicyCardsModule  viewState={viewState}
      updateViewState={updateViewState}
      policies={filteredPolicies}
      activeFilters={activeFilters}
      showPolicyCard={false}
      handleDrillDownClick={handleDrillDownClick} /></BrowserRouter>)
    expect(component).toMatchSnapshot()
  })
})
