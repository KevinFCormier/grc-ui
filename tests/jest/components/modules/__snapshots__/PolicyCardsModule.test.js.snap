// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PolicyCardsModule categories view renders collapse as expected 1`] = `
<div
  className="module-policy-cards"
>
  <div
    className="header-container"
  >
    <button
      className="collapse"
      onClick={[Function]}
    >
      <span
        className="collapse-hint"
      >
        Want to see your summary cards?
      </span>
      <span
        className="collapse-button"
      >
        Expand summary
      </span>
      <svg
        alt="expand"
        aria-label="expand"
        className="arrow-down"
        fill={undefined}
        fillRule="evenodd"
        height="7"
        name="chevron--down"
        role="img"
        style={undefined}
        viewBox="0 0 12 7"
        width="12"
      >
        <title>
          expand
        </title>
        <path
          d="M6.002 5.55L11.27 0l.726.685L6.003 7 0 .685.726 0z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`PolicyCardsModule categories view renders expand as expected 1`] = `
<div
  className="module-policy-cards"
>
  <div
    className="header-container"
  >
    <div
      className="header-title"
    >
      Summary
    </div>
    <div>
      <div
        aria-invalid={undefined}
        aria-label="Summary"
        className="bx--dropdown selection bx--list-box"
        data-invalid={undefined}
        onClick={[Function]}
        onKeyDown={[Function]}
        role="listbox"
        tabIndex={null}
      >
        <div
          aria-expanded={false}
          aria-haspopup={true}
          aria-label="open menu"
          className="bx--list-box__field"
          data-toggle={true}
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex="0"
          type="button"
        >
          <span
            className="bx--list-box__label"
            htmlFor="downshift-0-input"
          >
            Standards
          </span>
          <div
            className="bx--list-box__menu-icon"
          >
            <svg
              alt="Open menu"
              aria-label="Open menu"
              className={undefined}
              fill={undefined}
              fillRule="evenodd"
              focusable="false"
              height="5"
              name={undefined}
              role="img"
              style={undefined}
              viewBox="0 0 10 5"
              width="10"
            >
              <title>
                Open menu
              </title>
              <path
                d="M0 0l5 4.998L10 0z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <button
      className="collapse"
      onClick={[Function]}
    >
      <span
        className="collapse-hint"
      >
        Want to see less information?
      </span>
      <span
        className="collapse-button"
      >
        Collapse summary
      </span>
      <svg
        alt="collapse"
        aria-label="collapse"
        className="arrow-up"
        fill={undefined}
        fillRule="evenodd"
        height="7"
        name="chevron--up"
        role="img"
        style={undefined}
        viewBox="0 0 12 7"
        width="12"
      >
        <title>
          collapse
        </title>
        <path
          d="M5.994 1.45L.726 7 0 6.315 5.994 0l6.002 6.315L11.27 7z"
        />
      </svg>
    </button>
  </div>
  <div
    className="card-container-container"
  >
    <div>
      <div
        className="card-container"
      >
        <div
          className="card-content"
        >
          <div
            className="card-name"
          >
            NIST
          </div>
          <div
            className="card-count-content"
          >
            <div
              className="card-count-container violated"
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex="0"
            >
              <div
                className="card-count"
              >
                <div
                  className="card-count-violations"
                >
                  2
                </div>
                <div
                  className="card-count-total"
                >
                  /2
                </div>
              </div>
              <div
                className="card-violation-type"
              >
                Cluster violations
              </div>
            </div>
            <div
              className="card-count-container violated"
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex="0"
            >
              <div
                className="card-count"
              >
                <div
                  className="card-count-violations"
                >
                  2
                </div>
                <div
                  className="card-count-total"
                >
                  /2
                </div>
              </div>
              <div
                className="card-violation-type"
              >
                Policy violations
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        className="card-container"
      >
        <div
          className="card-content"
        >
          <div
            className="card-name"
          >
            HIPAA
          </div>
          <div
            className="card-count-content"
          >
            <div
              className="card-count-container violated"
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex="0"
            >
              <div
                className="card-count"
              >
                <div
                  className="card-count-violations"
                >
                  2
                </div>
                <div
                  className="card-count-total"
                >
                  /2
                </div>
              </div>
              <div
                className="card-violation-type"
              >
                Cluster violations
              </div>
            </div>
            <div
              className="card-count-container violated"
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex="0"
            >
              <div
                className="card-count"
              >
                <div
                  className="card-count-violations"
                >
                  1
                </div>
                <div
                  className="card-count-total"
                >
                  /1
                </div>
              </div>
              <div
                className="card-violation-type"
              >
                Policy violations
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        className="card-container"
      >
        <div
          className="card-content"
        >
          <div
            className="card-name"
          >
            Other
          </div>
          <div
            className="card-count-content"
          >
            <div
              className="card-count-container violated"
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex="0"
            >
              <div
                className="card-count"
              >
                <div
                  className="card-count-violations"
                >
                  2
                </div>
                <div
                  className="card-count-total"
                >
                  /2
                </div>
              </div>
              <div
                className="card-violation-type"
              >
                Cluster violations
              </div>
            </div>
            <div
              className="card-count-container violated"
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex="0"
            >
              <div
                className="card-count"
              >
                <div
                  className="card-count-violations"
                >
                  1
                </div>
                <div
                  className="card-count-total"
                >
                  /1
                </div>
              </div>
              <div
                className="card-violation-type"
              >
                Policy violations
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PolicyCardsModule standards view renders collapse as expected 1`] = `
<div
  className="module-policy-cards"
>
  <div
    className="header-container"
  >
    <button
      className="collapse"
      onClick={[Function]}
    >
      <span
        className="collapse-hint"
      >
        Want to see your summary cards?
      </span>
      <span
        className="collapse-button"
      >
        Expand summary
      </span>
      <svg
        alt="expand"
        aria-label="expand"
        className="arrow-down"
        fill={undefined}
        fillRule="evenodd"
        height="7"
        name="chevron--down"
        role="img"
        style={undefined}
        viewBox="0 0 12 7"
        width="12"
      >
        <title>
          expand
        </title>
        <path
          d="M6.002 5.55L11.27 0l.726.685L6.003 7 0 .685.726 0z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`PolicyCardsModule standards view renders expand as expected 1`] = `
<div
  className="module-policy-cards"
>
  <div
    className="header-container"
  >
    <div
      className="header-title"
    >
      Summary
    </div>
    <div>
      <div
        aria-invalid={undefined}
        aria-label="Summary"
        className="bx--dropdown selection bx--list-box"
        data-invalid={undefined}
        onClick={[Function]}
        onKeyDown={[Function]}
        role="listbox"
        tabIndex={null}
      >
        <div
          aria-expanded={false}
          aria-haspopup={true}
          aria-label="open menu"
          className="bx--list-box__field"
          data-toggle={true}
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex="0"
          type="button"
        >
          <span
            className="bx--list-box__label"
            htmlFor="downshift-1-input"
          >
            Standards
          </span>
          <div
            className="bx--list-box__menu-icon"
          >
            <svg
              alt="Open menu"
              aria-label="Open menu"
              className={undefined}
              fill={undefined}
              fillRule="evenodd"
              focusable="false"
              height="5"
              name={undefined}
              role="img"
              style={undefined}
              viewBox="0 0 10 5"
              width="10"
            >
              <title>
                Open menu
              </title>
              <path
                d="M0 0l5 4.998L10 0z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <button
      className="collapse"
      onClick={[Function]}
    >
      <span
        className="collapse-hint"
      >
        Want to see less information?
      </span>
      <span
        className="collapse-button"
      >
        Collapse summary
      </span>
      <svg
        alt="collapse"
        aria-label="collapse"
        className="arrow-up"
        fill={undefined}
        fillRule="evenodd"
        height="7"
        name="chevron--up"
        role="img"
        style={undefined}
        viewBox="0 0 12 7"
        width="12"
      >
        <title>
          collapse
        </title>
        <path
          d="M5.994 1.45L.726 7 0 6.315 5.994 0l6.002 6.315L11.27 7z"
        />
      </svg>
    </button>
  </div>
  <div
    className="card-container-container"
  >
    <div>
      <div
        className="card-container"
      >
        <div
          className="card-content"
        >
          <div
            className="card-name"
          >
            NIST
          </div>
          <div
            className="card-count-content"
          >
            <div
              className="card-count-container violated"
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex="0"
            >
              <div
                className="card-count"
              >
                <div
                  className="card-count-violations"
                >
                  2
                </div>
                <div
                  className="card-count-total"
                >
                  /2
                </div>
              </div>
              <div
                className="card-violation-type"
              >
                Cluster violations
              </div>
            </div>
            <div
              className="card-count-container violated"
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex="0"
            >
              <div
                className="card-count"
              >
                <div
                  className="card-count-violations"
                >
                  3
                </div>
                <div
                  className="card-count-total"
                >
                  /4
                </div>
              </div>
              <div
                className="card-violation-type"
              >
                Policy violations
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        className="card-container"
      >
        <div
          className="card-content"
        >
          <div
            className="card-name"
          >
            HIPAA
          </div>
          <div
            className="card-count-content"
          >
            <div
              className="card-count-container violated"
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex="0"
            >
              <div
                className="card-count"
              >
                <div
                  className="card-count-violations"
                >
                  2
                </div>
                <div
                  className="card-count-total"
                >
                  /2
                </div>
              </div>
              <div
                className="card-violation-type"
              >
                Cluster violations
              </div>
            </div>
            <div
              className="card-count-container violated"
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex="0"
            >
              <div
                className="card-count"
              >
                <div
                  className="card-count-violations"
                >
                  1
                </div>
                <div
                  className="card-count-total"
                >
                  /1
                </div>
              </div>
              <div
                className="card-violation-type"
              >
                Policy violations
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
